src/main/scala/workflow/RuleExecutor.scala

```diff
@@ -0,0 +1,85 @@
+package workflow
+
+import pipelines.Logging
+
+abstract class RuleExecutor extends Logging {
```

Comment:
I'm not using the Catalyst one directly here because it is parameterized on TreeNode[_], which pipelines don't really make sense as. It may be possible to force all Pipelines (including Transformers) to extend TreeNode[Pipeline] so we can use the built-in catalyst one (it's unclear, would need to dig into it a little bit).